### PR TITLE
RSDK-7526, RSDK-7527 - Fix deadlock in graph node

### PR DESCRIPTION
**Pull request by cheukt**
Cause
* the remote is removed from the config -> we call remove resource which eventually calls graphNode.Close which takes a lock on the graph node. The remote client Close gets stuck on waiting on activeBackgroundWorkers
* the active bg worker we're interested in is trying to reconnect to the remote client. And succeeds, somehow, and so tries to notifyParent. notifyParent involves pushing a message through the triggerConfig channel, and it gets stuck there
* the worker working on the triggerConfig messages is in the process of working through a previous reconfigure, and gets stuck in SaveSnapshot because it's waiting on a RLock on the graph node somewhere in exportNode. SaveSnapshot also took a graph lock at the beginning, so any other calls into the graph will lock (ResourceNames for one)

Fixes
* We should move the NeedsReconfigure() check to after we figure out whether the resource is a remote or not
    * Won’t fix main deadlock, but remove a thread
* graphNode.Close shouldn’t lock on the whole method but just the unassignment
    * After this unassignment, the graphNode doesn’t need to wait for Close to prevent any data races
* Add a buffer to triggerConfig and only send message to triggerConfig channel if it is not full
* Expanded usage of configLock to include deletions from the graph as well, and renamed it resourceGraphLock
    * idea would be that graphNode.Close would only be called within a resourceGraphLock, and since SaveSnapshot is as well, it will prevent deadlocks

TODO

- [ ] Add tests

### CODE REVIEW

**Comment by cheukt** [comment]
Wasn't able to get this into PR, but would like to get some of your thoughts around this patch @benjirewis @maximpertsov before I try to add tests. There is an alternate patch where we could not add another instance of configLock and instead take a shorter graph lock in ExportDot (grab the lock, make a copy of the graph, release lock), and just live with the chance that the resource nodes may change while we're iterating through the copy of the graph (which is possible with current main as well )

File: robot/client/client.go [hunk 1]
```diff
@@ -330,9 +330,10 @@ func (rc *RobotClient) connect(ctx context.Context) error {
 	if err := rc.connectWithLock(ctx); err != nil {
 		return err
 	}
-
+	rc.Logger().CInfow(ctx, "successfully reconnected to remote at address", "address", rc.address)
```
Comment (benjirewis) [inline robot/client/client.go:333]:
```suggestion
	rc.Logger().CInfow(ctx, "successfully (re)connected to remote at address", "address", rc.address)
```
This will also get logged during initial connection establishment, no?

File: robot/client/client.go [hunk 2]
```diff
@@ -330,9 +330,10 @@ func (rc *RobotClient) connect(ctx context.Context) error {
...
 	if rc.notifyParent != nil {
 		rc.notifyParent()
+		rc.Logger().CInfow(ctx, "successfully notified parent after reconnection", "address", rc.address)
```
Comment (benjirewis) [inline robot/client/client.go:336]: I sense this "parent" terminology will be confusing for most users. When exactly does this get logged? When an RDK remote connection or reconnection successfully sends to the `triggerConfig` channel or encounters a `closeContext` error? I would vote to not have this log (or have it at the debug level) unless you feel that connecting + being unable to notify parent is still a situation that exists after these changes.

File: robot/impl/resource_manager.go [hunk 3]
```diff
@@ -742,12 +748,17 @@ func (manager *resourceManager) completeConfigForRemotes(ctx context.Context, lr
...
+				// Attempt to trigger completeConfig goroutine execution when a change in remote
+				// is detected, but does not block if triggerConfig is full.
 				select {
 				case <-lr.closeContext.Done():
 					return
 				case lr.triggerConfig <- struct{}{}:
+				default:
+					rr.Logger().Debugw(
+						"remote attempted to trigger reconfiguration, but there is already one queued.",
+						"remote", remConf.Name,
```
Comment (benjirewis) [inline robot/impl/resource_manager.go:760]: [nit] Do you need to include remote name if we're already logging through the remote's client? Also, can we use `CDebugw`?

**Review by benjirewis — COMMENTED** [review]
Wow great find 👁️ . Just some nits. Were you able to reproduce the situation outlined in the PR description? If so, and it was reliably fixed with these changes, then I think you would only need a test for `graphNode.Close` only locking for unsetting `w.current`.

**Comment by benjirewis** [comment]
> just live with the chance that the resource nodes may change while we're iterating through the copy of the graph

Even if that possibility exists with your current code, I don't view it as a big deal. What's your opinion @maximpertsov? In any case, it actually feels more correct to me to add another `resourceGraphLock.Lock()` in `updateRemoteResourceNames` as you've done rather than lock the graph for a shorter period in `ExportDot`.

File: robot/impl/resource_manager.go [hunk 4]
```diff
@@ -51,8 +51,10 @@ type resourceManager struct {
 	moduleManager  modif.ModuleManager
 	opts           resourceManagerOptions
 	logger         logging.Logger
-	configLock     sync.Mutex
-	viz            resource.Visualizer
+
+	// resourceGraphLock manages access to the resource graph and nodes. If either may change, this lock should be taken.
+	resourceGraphLock sync.Mutex
```
Comment (maximpertsov) [inline robot/impl/resource_manager.go:56]: nice rename + clarifying comment - i always found `configLock` to be a bit confusing

File: robot/impl/local_robot.go [hunk 5]
```diff
@@ -382,11 +382,13 @@ func newWithResources(
...
-		closeContext:               closeCtx,
-		cancelBackgroundWorkers:    cancel,
-		triggerConfig:              make(chan struct{}),
+		operations:              operation.NewManager(logger),
+		logger:                  logger,
+		closeContext:            closeCtx,
+		cancelBackgroundWorkers: cancel,
+		// triggerConfig buffers 1 message so that we can queue up to 1 reconfiguration attempt
+		// (as long as there is 1 queued, further messages can be safely discarded).
+		triggerConfig:              make(chan struct{}, 1),
```
Comment (maximpertsov) [inline robot/impl/local_robot.go:391]: clever! deboucing is a nice improvement here

**Review by maximpertsov — COMMENTED** [review]
nice changes - I left some comments, but overall like the direction of this

File: robot/client/client.go (hunk 2 above)
Comment (cheukt) [inline robot/client/client.go:336]:
```suggestion
		rc.Logger().CDebugw(ctx, "successfully notified parent after reconnection", "remote address", rc.address)
```

File: robot/client/client.go [hunk 6]
```diff
@@ -716,7 +716,7 @@ func (rc *RobotClient) PackageManager() packages.Manager {
 //	resource_names := machine.ResourceNames()
 func (rc *RobotClient) ResourceNames() []resource.Name {
 	if err := rc.checkConnected(); err != nil {
-		rc.Logger().Errorw("failed to get remote resource names", "error", err)
+		rc.Logger().Errorw("failed to get remote resource names", "error", err.Error())
```
Comment (cheukt) [inline robot/client/client.go:719]: shortens the output from
```
2024-06-17T15:59:21.623Z        ERROR   robot_server.rdk:remote:/robot1 client/client.go:719    failed to get remote resource names       {"error":"not connected to remote robot at localhost:8080","errorVerbose":"not connected to remote robot at localhost:8080\ngo.viam.com/rdk/robot/client.(*RobotClient).notConnectedToRemoteError\n\trdk/robot/client/client.go:133\ngo.viam.com/rdk/robot/client.(*RobotClient).checkConnected\n\trdk/robot/client/client.go:492\ngo.viam.com/rdk/robot/client.(*RobotClient).ResourceNames\n\trdk/robot/client/client.go:718\ngo.viam.com/rdk/robot/impl.(*resourceManager).updateRemoteResourceNames\n\trdk/robot/impl/resource_manager.go:182\ngo.viam.com/rdk/robot/impl.(*resourceManager).updateRemotesResourceNames\n\trdk/robot/impl/resource_manager.go:280\ngo.viam.com/rdk/robot/impl.newWithResources.func3\n\trdk/robot/impl/local_robot.go:512\ngo.viam.com/utils.ManagedGo.func1\n\t/User/go/pkg/mod/go.viam.com/utils@v0.1.82/runtime.go:180\ngo.viam.com/utils.PanicCapturingGoWithCallback.func1\n\t/User/go/pkg/mod/go.viam.com/utils@v0.1.82/runtime.go:164\nruntime.goexit\n\t/opt/homebrew/Cellar/go@1.21/1.21.8/libexec/src/runtime/asm_arm64.s:1197"}
```
to
```
2024-06-17T15:56:45.315Z        ERROR   robot_server.rdk:remote:/robot1 client/client.go:719    failed to get remote resource names       {"error":"not connected to remote robot at localhost:8080"}
```

File: robot/impl/resource_manager.go [hunk 7]
```diff
@@ -449,7 +454,8 @@ func (manager *resourceManager) closeResource(ctx context.Context, res resource.
 	return allErrs
 }
 
-// closeAndUnsetResource attempts to close and unset the resource from the graph node.
+// closeAndUnsetResource attempts to close and unset the resource from the graph node. Should only be called within
+// resourceGraphLock.
```
[the remark this answers is not in the document]
Comment (cheukt) [reply robot/impl/resource_manager.go:458]: I'm not sure if I see closeAndUnsetResource grabbing a resourceGraphLock inside the function

File: robot/impl/resource_manager.go (hunk 3 above)
Comment (cheukt) [reply robot/impl/resource_manager.go]: updated

File: resource/graph_node_test.go [hunk 8]
```diff
@@ -213,13 +213,41 @@ func lifecycleTest(t *testing.T, node *resource.GraphNode, initialDeps []string)
 type someResource struct {
 	resource.Resource
 	closeCap  []interface{}
-	shoudlErr bool
+	shouldErr bool
```
Comment (cheukt) [inline resource/graph_node_test.go:216]: driveby spelling fix

**Comment by cheukt** [comment]
I don't really like that my additional test's failure condition is a deadlock, but also not sure if there's another good way to ensure correctness

File: robot/impl/resource_manager.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (maximpertsov) [reply robot/impl/resource_manager.go:458]: ugh, you're right, I misread the diff - we lock the function just below (`removeMarkedAndClose`) - it seemed like it was part of the same function and I overlooked the hint in the page break. ignore my comment.

**Comment by maximpertsov** [comment]
> I don't really like that my additional tests' failure condition is a deadlock, but also not sure if there's another good way to ensure correctness

Which new test causes the deadlock on fail? `TestSendTriggerConfig` ?

**Comment by maximpertsov** [comment]
> > just live with the chance that the resource nodes may change while we're iterating through the copy of the graph
> 
> Even if that possibility exists with your current code, I don't view it as a big deal. What's your opinion @maximpertsov? In any case, it actually feels more correct to me to add another `resourceGraphLock.Lock()` in `updateRemoteResourceNames` as you've done rather than lock the graph for a shorter period in `ExportDot`.

yep, this fix is superior and having inconsistent graph exports is not a big deal - especially if it's causing deadlocks in application code. the resource graph is still very much a "best-effort" and we can revisit it's consistency guarantees as we harden it. it might be worth adding a small comment about in `ExportDot`'s docstring to remind us that we explicitly made this decision.

**Comment by cheukt** [comment]
> > I don't really like that my additional tests' failure condition is a deadlock, but also not sure if there's another good way to ensure correctness
> 
> Which new test causes the deadlock on fail? `TestSendTriggerConfig` ?

Both `TestSendTriggerConfig` and `TestClose` will deadlock on failure

File: robot/client/client.go (hunk 2 above)
Comment (cheukt) [reply robot/client/client.go]: updated to `rc.Logger().CDebugw(ctx, "successfully notified parent after (re)connection", "address", rc.address)`

File: resource/visualizer.go [hunk 9]
```diff
@@ -268,6 +270,9 @@ func (g *Graph) ExportDot() (string, error) {
 	// the same text output is produced. We desire this such that a diff in the output implies the
 	// resource graph has logically changed. To achieve this, we sort all of the inputs (nodes and
 	// edges) for a predictable iteration order.
+	//
+	// As graph nodes are not all locked during the duration of this function, nodes could change
+	// before it gets exported. This is known and expected as the graph output is "best-effort".
```
Comment (cheukt) [inline resource/visualizer.go:275]: added to highlight consistency is not guaranteed

File: resource/graph_node_test.go [hunk 10]
```diff
@@ -213,13 +213,41 @@ func lifecycleTest(t *testing.T, node *resource.GraphNode, initialDeps []string)
...
+// Close calls the injected Close or the real version.
+func (a *anotherResource) Close(ctx context.Context) error {
+	if a.CloseFunc == nil {
+		return errors.New("oops")
+	}
+	return a.CloseFunc(ctx)
+}
+
+func TestClose(t *testing.T) {
+	// Tests that Close does not deadlock by calling graphNode.Close() inside a resource Close().
```
Comment (benjirewis) [inline resource/graph_node_test.go:241]: [q] How does this mimic the case described in the PR description?

File: robot/impl/local_robot_test.go [hunk 11]
```diff
@@ -3482,3 +3482,21 @@ func TestCustomResourceBuildsOnModuleAddition(t *testing.T) {
...
+
+func TestSendTriggerConfig(t *testing.T) {
+	// Tests that the triggerConfig channel buffers exactly 1 message and that
+	// sendTriggerConfig does not block whether or not the channel has
+	// capacity.
+	ctx := context.Background()
+	logger := logging.NewTestLogger(t)
+
+	r := setupLocalRobot(t, ctx, &config.Config{}, logger)
+	r.Close(ctx)
```
Comment (maximpertsov) [inline robot/impl/local_robot_test.go:3494]: why do we need to close here?

Comment (benjirewis) [inline robot/impl/local_robot_test.go:3494]: Why are we `Close`ing immediately in this test? Won't that cause `sendTriggerConfig` to be a noop (surprised this test passes)?

Comment (benjirewis) [reply robot/impl/local_robot_test.go:3494]: Oops @maximpertsov asked before me 😆 .

Comment (cheukt) [reply robot/impl/local_robot_test.go:3494]: the close stops the reconfigureWorker, so closing it here means that no messages will be processed. the test is checking that sendTriggerConfig is attempting to send messages to the channel but not blocking if it fails. That's why there's a check that there is exactly one message in the `triggerConfig` channel

Comment (cheukt) [reply robot/impl/local_robot_test.go:3494]: answered here https://github.com/viamrobotics/rdk/pull/3968#discussion_r1643314785

File: resource/graph_node_test.go (hunk 10 above)
Comment (cheukt) [reply resource/graph_node_test.go:241]: this test is a bit narrower - this only tests that graphNode.Close() does not hold the lock while Closing the resource. In the PR description, the issue was that graphNode.Close() held the lock while Closing the resource, and the Close ended up depending on something else that wanted the graphNode lock

File: robot/impl/local_robot_test.go (hunk 11 above)
Comment (benjirewis) [reply robot/impl/local_robot_test.go:3494]: Won't this piece of `sendTriggerConfig` stop _any_ of the 5 calls from sending anything to the channel?
```
if r.closeContext.Err() != nil {
	return
}
```

File: resource/graph_node_test.go (hunk 10 above)
Comment (benjirewis) [reply resource/graph_node_test.go:241]: Cool that makes sense; thanks!

File: robot/impl/local_robot_test.go (hunk 11 above)
Comment (cheukt) [reply robot/impl/local_robot_test.go:3494]: that's why I re-initialize the context [here](https://github.com/viamrobotics/rdk/pull/3968/files#diff-d44b43f9f1e7b665af6465a60419c6684582f59225bcde5e60d7931932895c60R3497).

Comment (benjirewis) [reply robot/impl/local_robot_test.go]: Oops totally missed that; thanks.

Comment (cheukt) [reply robot/impl/local_robot_test.go]: also added more comments in the test

File: robot/impl/local_robot_test.go [hunk 12]
```diff
@@ -3482,3 +3482,26 @@ func TestCustomResourceBuildsOnModuleAddition(t *testing.T) {
...
+	ctx := context.Background()
+	logger := logging.NewTestLogger(t)
+
+	// Set up local robot normally so that the triggerConfig channel is set up normally
+	r := setupLocalRobot(t, ctx, &config.Config{}, logger)
+
+	// Close the robot to stop the background workers from processing any messages to triggerConfig
+	// but also reinitialize the closeContext so that sendTriggerConfig will attempt to send messages
+	// through.
+	r.Close(ctx)
```
Comment (maximpertsov) [inline robot/impl/local_robot_test.go:3499]: [minor] we expect to this to not error right?

```suggestion
	test.That(t, r.Close(ctx), test.ShouldBeNil)
```

Comment (maximpertsov) [reply robot/impl/local_robot_test.go:3499]: or more importantly, would there be an issue if there were an error?

Comment (cheukt) [reply robot/impl/local_robot_test.go]: I suppose so, that may mean that the closeContext did not get closed properly

File: resource/graph_node_test.go [hunk 13]
```diff
@@ -213,13 +213,41 @@ func lifecycleTest(t *testing.T, node *resource.GraphNode, initialDeps []string)
...
+	node := resource.NewConfiguredGraphNode(
+		resource.Config{},
+		ourRes,
+		resource.DefaultModelFamily.WithModel("bar"),
+	)
+
+	ourRes.CloseFunc = func(ctx context.Context) error {
+		return node.Close(ctx)
+	}
+	test.That(t, node.Close(context.Background()), test.ShouldBeNil)
```
[the remark this answers is not in the document]
Comment (cheukt) [reply resource/graph_node_test.go]: I think 20s should in theory be enough? for the resource manager test, it takes 2s normally and for this test it's like 0.75s

[the remark this answers is not in the document]
Comment (maximpertsov) [reply resource/graph_node_test.go]: yeah, I don't know what the right threshold is to avoid time-based flakes - but even something on higher side is better than a deadlock.

**Review by maximpertsov — APPROVED** [review]

File: resource/graph_node_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (maximpertsov) [reply resource/graph_node_test.go]: ah you did already, great!
